// @devnote-dev/pterojs/dist/src/application/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    users: {
        main: '/users',
        get: (u) => `/users/${u}`,
        ext: (u) => `/users/external/${u}`
    },
    nodes: {
        main: '/nodes',
        get: (n) => `/nodes/${n}`,
        deploy: '/nodes/deployable',
        config: (n) => `/nodes/${n}/configuration`,
        allocations: {
            main: (n) => `/nodes/${n}/allocations`,
            get: (n, a) => `/nodes/${n}/allocations/${a}`
        }
    },
    servers: {
        main: '/servers',
        get: (s) => `/servers/${s}`,
        ext: (s) => `/servers/external/${s}`,
        details: (s) => `/servers/${s}/details`,
        build: (s) => `/servers/${s}/build`,
        startup: (s) => `/servers/${s}/startup`,
        suspend: (s) => `/servers/${s}/suspend`,
        unsuspend: (s) => `/servers/${s}/unsuspend`,
        reinstall: (s) => `/servers/${s}/reinstall`,
        databases: {
            main: (s) => `/servers/${s}/databases`,
            get: (s, id) => `/servers/${s}/databases/${id}`,
            reset: (s, id) => `/servers/${s}/databases/${id}/reset-password`
        }
    },
    locations: {
        main: '/locations',
        get: (l) => `/locations/${l}`
    },
    nests: {
        main: '/nests',
        get: (n) => `/nests/${n}`,
        eggs: {
            main: (n) => `/nests/${n}/eggs`,
            get: (n, e) => `/nests/${n}/eggs/${e}`
        }
    }
};
