// @devnote-dev/pterojs/dist/src/client/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    account: {
        main: '/account',
        tfa: '/account/two-factor',
        email: '/account/email',
        password: '/account/password',
        apikeys: {
            main: '/account/api-keys',
            get: (k) => `/account/api-keys/${k}`
        },
        activity: '/account/activity',
        sshkeys: {
            main: '/account/ssh-keys',
            remove: '/account/ssh-keys/remove'
        }
    },
    servers: {
        main: '',
        get: (s) => `/servers/${s}`,
        databases: {
            main: (s) => `/servers/${s}/databases`,
            get: (s, id) => `/servers/${s}/databases/${id}`,
            rotate: (s, id) => `/servers/${s}/databases/${id}/rotate-password`
        },
        files: {
            main: (s) => `/servers/${s}/files/list`,
            contents: (s, f) => `/servers/${s}/files/contents?file=${f}`,
            download: (s, f) => `/servers/${s}/files/download?file=${f}`,
            rename: (s) => `/servers/${s}/files/rename`,
            copy: (s) => `/servers/${s}/files/copy`,
            write: (s, f) => `/servers/${s}/files/write?file=${f}`,
            compress: (s) => `/servers/${s}/files/compress`,
            decompress: (s) => `/servers/${s}/files/decompress`,
            delete: (s) => `/servers/${s}/files/delete`,
            create: (s) => `/servers/${s}/files/create-folder`,
            upload: (s) => `/servers/${s}/files/upload`,
            chmod: (s) => `/servers/${s}/files/chmod`
        },
        schedules: {
            main: (s) => `/servers/${s}/schedules`,
            get: (s, id) => `/servers/${s}/schedules/${id}`,
            exec: (s, id) => `/servers/${s}/schedules/${id}/execute`,
            tasks: {
                main: (s, id) => `/servers/${s}/schedules/${id}/tasks`,
                get: (s, id, t) => `/servers/${s}/schedules/${id}/tasks/${t}`
            }
        },
        network: {
            main: (s) => `/servers/${s}/network/allocations`,
            get: (s, id) => `/servers/${s}/network/allocations/${id}`,
            primary: (s, id) => `/servers/${s}/network/allocations/${id}/primary`
        },
        users: {
            main: (s) => `/servers/${s}/users`,
            get: (s, id) => `/servers/${s}/users/${id}`
        },
        backups: {
            main: (s) => `/servers/${s}/backups`,
            get: (s, id) => `/servers/${s}/backups/${id}`,
            lock: (s, id) => `/servers/${s}/backups/${id}/lock`,
            download: (s, id) => `/servers/${s}/backups/${id}/download`,
            restore: (s, id) => `/servers/${s}/backups/${id}/restore`
        },
        startup: {
            get: (s) => `/servers/${s}/startup`,
            var: (s) => `/servers/${s}/startup/variable`
        },
        settings: {
            image: (s) => `/servers/${s}/settings/docker-image`,
            rename: (s) => `/servers/${s}/settings/rename`,
            reinstall: (s) => `/servers/${s}/settings/reinstall`
        },
        ws: (s) => `/servers/${s}/websocket`,
        resources: (s) => `/servers/${s}/resources`,
        command: (s) => `/servers/${s}/command`,
        power: (s) => `/servers/${s}/power`
    },
    permissions: '/permissions'
};
