// @devnote-dev/pterojs/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.configLoader = exports.caseConv = exports.escape = exports.UserBuilder = exports.ServerBuilder = exports.NodeBuilder = exports.Builder = exports.Schedule = exports.Node = exports.ClientServer = exports.BaseManager = exports.ApplicationServer = exports.WebSocketManager = exports.SubUserManager = exports.Shard = exports.ScheduleManager = exports.NetworkManager = exports.FileManager = exports.ClientDatabaseManager = exports.ClientServerManager = exports.BackupManager = exports.PteroClient = exports.UserManager = exports.NodeManager = exports.NodeLocationManager = exports.NodeAllocationManager = exports.NestManager = exports.NestEggsManager = exports.ApplicationServerManager = exports.ApplicationDatabaseManager = exports.PteroApp = exports.version = void 0;
const package_json_1 = require("../package.json");
exports.version = package_json_1.version;
// Application API
var application_1 = require("./application");
Object.defineProperty(exports, "PteroApp", { enumerable: true, get: function () { return application_1.PteroApp; } });
var ApplicationDatabaseManager_1 = require("./application/ApplicationDatabaseManager");
Object.defineProperty(exports, "ApplicationDatabaseManager", { enumerable: true, get: function () { return ApplicationDatabaseManager_1.ApplicationDatabaseManager; } });
var ApplicationServerManager_1 = require("./application/ApplicationServerManager");
Object.defineProperty(exports, "ApplicationServerManager", { enumerable: true, get: function () { return ApplicationServerManager_1.ApplicationServerManager; } });
var NestEggsManager_1 = require("./application/NestEggsManager");
Object.defineProperty(exports, "NestEggsManager", { enumerable: true, get: function () { return NestEggsManager_1.NestEggsManager; } });
var NestManager_1 = require("./application/NestManager");
Object.defineProperty(exports, "NestManager", { enumerable: true, get: function () { return NestManager_1.NestManager; } });
var NodeAllocationManager_1 = require("./application/NodeAllocationManager");
Object.defineProperty(exports, "NodeAllocationManager", { enumerable: true, get: function () { return NodeAllocationManager_1.NodeAllocationManager; } });
var NodeLocationManager_1 = require("./application/NodeLocationManager");
Object.defineProperty(exports, "NodeLocationManager", { enumerable: true, get: function () { return NodeLocationManager_1.NodeLocationManager; } });
var NodeManager_1 = require("./application/NodeManager");
Object.defineProperty(exports, "NodeManager", { enumerable: true, get: function () { return NodeManager_1.NodeManager; } });
var UserManager_1 = require("./application/UserManager");
Object.defineProperty(exports, "UserManager", { enumerable: true, get: function () { return UserManager_1.UserManager; } });
// Client API
var client_1 = require("./client");
Object.defineProperty(exports, "PteroClient", { enumerable: true, get: function () { return client_1.PteroClient; } });
var BackupManager_1 = require("./client/BackupManager");
Object.defineProperty(exports, "BackupManager", { enumerable: true, get: function () { return BackupManager_1.BackupManager; } });
var ClientServerManager_1 = require("./client/ClientServerManager");
Object.defineProperty(exports, "ClientServerManager", { enumerable: true, get: function () { return ClientServerManager_1.ClientServerManager; } });
var ClientDatabaseManager_1 = require("./client/ClientDatabaseManager");
Object.defineProperty(exports, "ClientDatabaseManager", { enumerable: true, get: function () { return ClientDatabaseManager_1.ClientDatabaseManager; } });
var FileManager_1 = require("./client/FileManager");
Object.defineProperty(exports, "FileManager", { enumerable: true, get: function () { return FileManager_1.FileManager; } });
var NetworkManager_1 = require("./client/NetworkManager");
Object.defineProperty(exports, "NetworkManager", { enumerable: true, get: function () { return NetworkManager_1.NetworkManager; } });
var ScheduleManager_1 = require("./client/ScheduleManager");
Object.defineProperty(exports, "ScheduleManager", { enumerable: true, get: function () { return ScheduleManager_1.ScheduleManager; } });
var Shard_1 = require("./client/ws/Shard");
Object.defineProperty(exports, "Shard", { enumerable: true, get: function () { return Shard_1.Shard; } });
var SubUserManager_1 = require("./client/SubUserManager");
Object.defineProperty(exports, "SubUserManager", { enumerable: true, get: function () { return SubUserManager_1.SubUserManager; } });
var WebSocketManager_1 = require("./client/ws/WebSocketManager");
Object.defineProperty(exports, "WebSocketManager", { enumerable: true, get: function () { return WebSocketManager_1.WebSocketManager; } });
// Commons
__exportStar(require("./common"), exports);
__exportStar(require("./common/app"), exports);
__exportStar(require("./common/client"), exports);
// HTTP
__exportStar(require("./http/RequestManager"), exports);
// Structures
var ApplicationServer_1 = require("./structures/ApplicationServer");
Object.defineProperty(exports, "ApplicationServer", { enumerable: true, get: function () { return ApplicationServer_1.ApplicationServer; } });
var BaseManager_1 = require("./structures/BaseManager");
Object.defineProperty(exports, "BaseManager", { enumerable: true, get: function () { return BaseManager_1.BaseManager; } });
var ClientServer_1 = require("./structures/ClientServer");
Object.defineProperty(exports, "ClientServer", { enumerable: true, get: function () { return ClientServer_1.ClientServer; } });
__exportStar(require("./structures/Dict"), exports);
__exportStar(require("./structures/Errors"), exports);
var Node_1 = require("./structures/Node");
Object.defineProperty(exports, "Node", { enumerable: true, get: function () { return Node_1.Node; } });
__exportStar(require("./structures/Permissions"), exports);
var Schedule_1 = require("./structures/Schedule");
Object.defineProperty(exports, "Schedule", { enumerable: true, get: function () { return Schedule_1.Schedule; } });
__exportStar(require("./structures/User"), exports);
// Builders
var base_1 = require("./builders/base");
Object.defineProperty(exports, "Builder", { enumerable: true, get: function () { return base_1.Builder; } });
var Node_2 = require("./builders/Node");
Object.defineProperty(exports, "NodeBuilder", { enumerable: true, get: function () { return Node_2.NodeBuilder; } });
var Server_1 = require("./builders/Server");
Object.defineProperty(exports, "ServerBuilder", { enumerable: true, get: function () { return Server_1.ServerBuilder; } });
var User_1 = require("./builders/User");
Object.defineProperty(exports, "UserBuilder", { enumerable: true, get: function () { return User_1.UserBuilder; } });
// Utilities
var escape_1 = require("./util/escape");
Object.defineProperty(exports, "escape", { enumerable: true, get: function () { return escape_1.default; } });
var caseConv_1 = require("./util/caseConv");
Object.defineProperty(exports, "caseConv", { enumerable: true, get: function () { return caseConv_1.default; } });
var config_1 = require("./util/config");
Object.defineProperty(exports, "configLoader", { enumerable: true, get: function () { return config_1.default; } });
__exportStar(require("./util/query"), exports);
